src/tools/tests/schema2kotlin-test.ts

```diff
@@ -219,10 +250,123 @@ describe('schema2kotlin', () => {
         'typealias P_H1_2 = AbstractP.P_H1_2',
       ]
     ));
+    it('Progressively constrained variables', async () => await assertSchemaAliases(
+      `particle T
+         h1: reads ~a
+         h2: writes ~a with {amt: Number}
+         h3: reads writes ~a with {name: Text, age: Number}
+      `, [
+        'typealias T_H1 = AbstractT.TInternal1',
+        'typealias T_H2 = AbstractT.TInternal1',
+        'typealias T_H3 = AbstractT.TInternal1',
+      ]
+    ));
     async function assertSchemaAliases(manifest: string, expectedAliases: string[]) {
       await assertComponent(manifest, ({typeAliases}) => typeAliases.sort(), expectedAliases);
     }
   });
+  describe('Kotlin Generator', () => {
+    it('generates entity with public constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates variable entity with private constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads ~a with {num: Number}
+         `,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1 private constructor(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates copy and mutate by entity fields', async () => await assertCopyMethods(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = T_H1(num = num)
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = T_H1(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        )`
+    ));
+    // TODO(alxr): Why do we omit the docstring for Wasm?
```

Comment:
Question for the reviewers.

Reply:
hahaha, I don't know 😄 Wasm devs are hackers and don't need docs.

Reply:
Lol :P 
So, shall I add it back?

Reply:
They weren't there. I didn't remove the docstring.